### ISSUE TITLE
simple Seraching

### ISSUE THREAD

**Issue by Ahbee** [nTopology/cgal_dDSearch#1]
Not coding right now but I was just going to do something simple like this 

mTree is the Orthogonal_k_neighbor_search tree with a float kernel
```c++
bool UVEditor::isDuplicate(const ci::Vec2f &point){
    auto query = Point_f{toCGAL(point)};
    Neighbor_search search(mTree, query, 1);
   auto distance = (search.begin())->second;
    return nt::epsilonEquals(distance, 0.f);
}
```

this code should been done right before you add the polygon
```c++
    if (!isDuplicate(center)) {
        // grow
        mTree.insert(toCGAL(center));
    }
```

```c++
NTKernel::Point_2 toCGAL(const ci::Vec2f& p){
    return NTKernel::Point_2{p.x,p.y};
}

```

I have not tested this code though

**Comment by bradrothenberg** [comment]
Do you know how to remove pts from the tree?

BRADLEY ROTHENBERG
Bradley Rothenberg | co-founder |  381-383 Broadway, #203 NY, NY 10013   
t. 917.710.1396   e. brad@bradleyrothenberg.com  w.bradleyrothenberg.com
  
twitter  l  instagram  l  facebook

**Comment by Ahbee** [comment]
I think the only way is to clear and rebuild the tree. rebuilding the tree is not that bad. Runs in real time for 3000 points at least for 2d. I am saving a differnet tree for each generation.

**Comment by Ahbee** [comment]
clocked it building the tree is fast.
rebuilding the tree for 2000 points takes  .65 ms on mac debug mode which is .00065 seconds.
```c++
    auto start = chrono::steady_clock::now();
    mTree.clear();
    for(auto & s :mShapes){
        auto sCenter = s->getPoints()[0]->getWorldPosition();
        mTree.insert(toCGAL(sCenter));
    }
    auto end = chrono::steady_clock::now();
    
    auto diff = end - start;
    ci::app::console() << chrono::duration <double, milli> (diff).count() << " ms" << endl;
```

**Comment by bradrothenberg** [comment]
ok wow, thats great.

So worst case we need to rebuild the tree whenever we remove pieces




*nTopology Inc.*

Bradley Rothenberg | co-founder |  381-383 Broadway, #203 NY, NY 10013

t. 917.710.1396   e. *brad@bradleyrothenberg.com
<brad@bradleyrothenberg.com>*  w. *nTopology.com <http://nTopology.com>*



*twitter*   <https://twitter.com/brad_rothenberg>l  *instagram
<http://instagram.com/bradleyrothenberg>  *l  *facebook*
<https://www.facebook.com/bradley.rothenberg>

**Comment by bradrothenberg** [comment]
did you have a chance to look at this file?

**Comment by Ahbee** [comment]
> did you have a chance to look at this file?

which file?

**Comment by bradrothenberg** [comment]
cgal_dDSearch

**Comment by Ahbee** [comment]
This repo? I just saw that you were trying to use custom types, which I felt was unnecessary because it is simple to convert cinder points to cgal points. I would rather create a custom kernel that works with all modules than create something specific for this module.

**Comment by bradrothenberg** [comment]
Yeah, but I couldn't figure out how to get the tree to work using cgals point_3 -- the only one I got to work was this module using 3D points -- thats why I wanted you to look at it.

I also tried to setup the tree like this:

```c++
typedef CGAL::Simple_cartesian<float> K;
typedef K::Point_3 Point_d;

typedef CGAL::Search_traits_3<K> TreeTraits;
typedef CGAL::Euclidean_distance <TreeTraits> distance;
```

& could not get the tree to work

**Comment by bradrothenberg** [comment]
I don't know what I need to typedef to setup the tree to work w/ cgals point_3

**Comment by bradrothenberg** [comment]
nm, figured it out:

```c++
typedef CGAL::Simple_cartesian<float> K;
typedef K::Point_3 Point_d;
typedef CGAL::Search_traits_3<K> TreeTraits;
typedef CGAL::Orthogonal_k_neighbor_search<TreeTraits> Neighbor_search;
typedef Neighbor_search::Tree Tree;
```
and
```c++
	const unsigned int N = 1;
	std::list<Point_d> points;
	points.push_back(Point_d(0, 0,0));
	Tree tree(points.begin(), points.end());
	Point_d query(0, 0,1);
	// Initialize the search structure, and search all N points
	Neighbor_search search(tree, query, N);
	// report the N nearest neighbors and their distance
	// This should sort all N points by increasing distance from origin
	for (Neighbor_search::iterator it = search.begin(); it != search.end(); ++it){
		ci::app::console() << it->first << " - dist:" << std::sqrt(it->second) << std::endl;
	}


```
works

**Comment by Ahbee** [comment]
I tried that but it does not work for me, but if I change the loop to this it works.
```c++
for (auto it = search.advanced_begin(); it != search.advanced_end(); ++it){
        ci::app::console() << *(it->first) << " - dist:" << std::sqrt(it->second) << std::endl;
    }
```

**Comment by Ahbee** [comment]
@bradrothenberg , for optimal performance we should really use range queries, since we really only need to search around a small radius around the point, to check for duplicates. I am looking into it now let me now if you figure it out before I do.

**Comment by Ahbee** [comment]
nm the range queries actually seem to be a lot slower
```c++
bool UVEditor::isDuplicate(const ci::Vec2f &point){
    if (mTree.size() == 0) {
        return false;
    }
    
    auto query = Point_f{toCGAL(point)};
    auto fuzzySphere = Fuzzy_sphere{query, .0001};
    
    auto search = std::vector<Point_f>{};
    
    mTree.search(std::back_inserter(search), fuzzySphere);
    
    if (search.size() > 0) {
        return true;
    }
    
    return false;
}

```

**Comment by bradrothenberg** [comment]
really?

That seems weird, shouldn't it be faster?

**Comment by Ahbee** [comment]
Not sure, I tried the both 3.3 and 3.4 styles they were both slow. 3.1 is the fastest.

but I am actually looking at the spatial sorting module 
http://doc.cgal.org/latest/Spatial_sorting/index.html#Chapter_Spatial_Sorting

this makes more sense to me because what we need is just to store all the centers in a sorted vector. And then we want to check if a point already exists in the vector. if it exists then we do not add the polygon. If it doesn't then we insert into the sorted list.

**Comment by bradrothenberg** [comment]
Do you think that will be faster? it looks like it might be

**Comment by Ahbee** [comment]
> Do you think that will be faster? it looks like it might be

dont know it depends on how fast the sorting is.

**Comment by bradrothenberg** [comment]
The CGAL spatial searching seems a lot slower than the grid search, if that's possible? It slows down to over 1.5 seconds to generate 326 new objects w/ 1000 pts in the tree.

that seems way too slow.

**Comment by bradrothenberg** [comment]
it takes 3.3s in release mode to generate 578 new objects out of (2916) w/ 2625 pts in the tree

**Comment by Ahbee** [comment]
> it takes 3.3s in release mode to generate 578 new objects out of (2916) w/ 2625 pts in the tree

Is this the kd tree, 3.1 style?

I have not tried the spatial sorting yet so no comment there. But the grid should be the fastest way, if you are implementing it right. Since it order 1 to place a point in a cell.

**Comment by bradrothenberg** [comment]
yeah exactly -- w/ Orthogonal_k_neighbor_search

**Comment by Ahbee** [comment]
what do mean 2916? the number of objects should always equal the points in the tree.

**Comment by bradrothenberg** [comment]
the 2916 is how many objects were tried to add, but only 578 were added because the rest were duplicates...the size of all objects is 2625 which is the tree size.

**Comment by Ahbee** [comment]
are you growing by generation? can you print how many new objects were added each generation.

**Comment by bradrothenberg** [comment]
Yeah, I'm doing that -- I'm checking to see if making he new objects is what is slow

**Comment by Ahbee** [comment]
Another thing to make sure of is that you are only inserting new points into the tree after all children are added, because there there is no way that children of the same shape can overlap.

**Comment by bradrothenberg** [comment]
After all children per shape? Or generation? Children w/ different parents but the same gen can overlap

**Comment by Ahbee** [comment]
Just the count of the current generation(or last generation) after each growth step.  so image the basic square rule , it would print 
```
1
4
8
```
